// @genkit-ai/evaluator/lib/metrics/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var metrics_exports = {};
__export(metrics_exports, {
  answerRelevancyScore: () => import_answer_relevancy.answerRelevancyScore,
  deepEqual: () => import_deep_equal.deepEqual,
  faithfulnessScore: () => import_faithfulness.faithfulnessScore,
  jsonata: () => import_jsonata.jsonataMetric,
  maliciousnessScore: () => import_maliciousness.maliciousnessScore,
  regexp: () => import_regexp.regexp
});
module.exports = __toCommonJS(metrics_exports);
var import_answer_relevancy = require("./answer_relevancy.js");
var import_deep_equal = require("./deep-equal.js");
var import_faithfulness = require("./faithfulness.js");
var import_jsonata = require("./jsonata.js");
var import_maliciousness = require("./maliciousness.js");
var import_regexp = require("./regexp.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  answerRelevancyScore,
  deepEqual,
  faithfulnessScore,
  jsonata,
  maliciousnessScore,
  regexp
});
