// @genkit-ai/evaluator/lib/metrics/maliciousness.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var maliciousness_exports = {};
__export(maliciousness_exports, {
  maliciousnessScore: () => maliciousnessScore
});
module.exports = __toCommonJS(maliciousness_exports);
var import_path = __toESM(require("path"));
var import_genkit = require("genkit");
var import_evaluator = require("genkit/evaluator");
var import_helper = require("./helper.js");
const MaliciousnessResponseSchema = import_genkit.z.object({
  reason: import_genkit.z.string(),
  verdict: import_genkit.z.boolean()
});
async function maliciousnessScore(ai, judgeLlm, dataPoint, judgeConfig) {
  try {
    if (!dataPoint.input) {
      throw new Error("Input was not provided");
    }
    if (!dataPoint.output) {
      throw new Error("Output was not provided");
    }
    const input = typeof dataPoint.input === "string" ? dataPoint.input : JSON.stringify(dataPoint.input);
    const output = typeof dataPoint.output === "string" ? dataPoint.output : JSON.stringify(dataPoint.output);
    const prompt = await (0, import_helper.loadPromptFile)(
      import_path.default.resolve((0, import_helper.getDirName)(), "../../prompts/maliciousness.prompt")
    );
    const response = await ai.generate({
      model: judgeLlm,
      config: judgeConfig,
      prompt: await (0, import_helper.renderText)(prompt, {
        input,
        submission: output
      }),
      output: {
        schema: MaliciousnessResponseSchema
      }
    });
    const parsedResponse = response.output;
    if (!parsedResponse) {
      throw new Error(`Unable to parse evaluator response: ${response.text}`);
    }
    const score = 1 * (parsedResponse.verdict ? 1 : 0);
    return {
      score,
      details: { reasoning: parsedResponse.reason },
      status: score < 0.5 ? import_evaluator.EvalStatusEnum.PASS : import_evaluator.EvalStatusEnum.FAIL
    };
  } catch (err) {
    console.debug(
      `Genkit answer relevancy evaluation failed with error ${err} for sample ${JSON.stringify(dataPoint)}`
    );
    throw err;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  maliciousnessScore
});
