// @genkit-ai/evaluator/lib/types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  GenkitMetric: () => GenkitMetric,
  isGenkitMetricConfig: () => isGenkitMetricConfig
});
module.exports = __toCommonJS(types_exports);
var GenkitMetric = /* @__PURE__ */ ((GenkitMetric2) => {
  GenkitMetric2["FAITHFULNESS"] = "FAITHFULNESS";
  GenkitMetric2["ANSWER_RELEVANCY"] = "ANSWER_RELEVANCY";
  GenkitMetric2["ANSWER_ACCURACY"] = "ANSWER_ACCURACY";
  GenkitMetric2["MALICIOUSNESS"] = "MALICIOUSNESS";
  GenkitMetric2["REGEX"] = "REGEX";
  GenkitMetric2["DEEP_EQUAL"] = "DEEP_EQUAL";
  GenkitMetric2["JSONATA"] = "JSONATA";
  return GenkitMetric2;
})(GenkitMetric || {});
function isGenkitMetricConfig(input) {
  return Object.hasOwn(input, "type");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  GenkitMetric,
  isGenkitMetricConfig
});
